AQLM quantizer support

BlackSamorez opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes https://github.com/Vahe1994/AQLM/issues/11


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@younesbelkada

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Commit:
aqlm init

Commit:
calibration and dtypes

Commit:
docs

BlackSamorez commented:
A model to test it: [BlackSamorez/Mixtral-8x7b-AQLM-2Bit-1x16-hf-test-dispatch](https://huggingface.co/BlackSamorez/Mixtral-8x7b-AQLM-2Bit-1x16-hf-test-dispatch)

BlackSamorez commented:
A Google Colab demo: [Mixtral in 2 bits](https://colab.research.google.com/drive/1-xZmBRXT5Fm3Ghn4Mwa2KRypORXb855X?usp=sharing).

younesbelkada commented on docs/source/en/quantization.md:
````diff
@@ -518,6 +518,19 @@ double_quant_config = BitsAndBytesConfig(
 model_double_quant = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-13b", quantization_config=double_quant_config)
 ```
 
+## AQLM
````
Would it be possible to move that on the top of the doc page before AWQ? we sort them in the integration order

BlackSamorez replied:
Done. I also added a short description of available inference setups.

younesbelkada replied:
Awesome thanks !

younesbelkada commented on src/transformers/quantizers/quantizer_aqlm.py:
```diff
@@ -0,0 +1,141 @@
+        self.quantization_config = quantization_config
+
+    def validate_environment(self, *args, **kwargs):
+        if not is_accelerate_available():
```
Could you also add a method in import_utils `is_aqlm_available()` and use that method here as well?

BlackSamorez replied:
Done

younesbelkada replied:
Nice, thanks !

younesbelkada commented on src/transformers/quantizers/quantizer_aqlm.py:
```diff
@@ -0,0 +1,141 @@
+
+    def update_torch_dtype(self, torch_dtype: "torch.dtype") -> "torch.dtype":
+        if torch_dtype is None:
+            if torch.cuda.is_available():
+                torch_dtype = torch.float16
+                logger.info(
+                    "CUDA available. Assuming AQLM inference on GPU and loading the model in `torch.float16`. To overwrite it, set `torch_dtype` manually."
+                )
+            else:
+                torch_dtype = torch.float32
+                logger.info(
+                    "CUDA is unavailable. Assuming AQLM inference on CPU and loading the model in `torch.float32`. To overwrite it, set `torch_dtype` manually."
+                )
```
Very nice !

BlackSamorez replied:
Thanks!

younesbelkada commented on src/transformers/quantizers/quantizer_aqlm.py:
```diff
@@ -0,0 +1,141 @@
+
+        if isinstance(module, nn.Linear):
+            # Check if the current key is not in the `linear_weights_not_to_quantize`
+            if ".".join(current_key_name) + ".weight" not in linear_weights_not_to_quantize:
```
I saw in the config of the model you pushed on the Hub that you also included layer norm weights inside `linear_weights_not_to_quantize` , I think these can be excluded from the config as they are not an insitance of `nn.Linear` right?

BlackSamorez replied:
They certainly can be excluded. It's just that converting from a freshly quantized AQLM format it would be troublesome to check if an unquantized `.weight` parameter is of `nn.Linear` or not. So I simply included all of them just in case. That Mixtral config can, indeed, be made somewhat shorter.

younesbelkada commented on docs/source/en/quantization.md:
```diff
@@ -518,6 +518,19 @@ double_quant_config = BitsAndBytesConfig(
+
+The library provides efficient kernels for both GPU and CPU inference.
+
+The instructions on how to quantize models yourself, as well as all the relevant code can be found in the corresponding GitHub [repository](https://github.com/Vahe1994/AQLM).
```
Can you also put the link to the google colab here?

BlackSamorez replied:
Done

younesbelkada replied:
Thanks !

younesbelkada reviewed (commented):
Looks pretty clean already ! Thanks so much for working on this and converting Mixtral using AQLM format - thanks also for sharing the Google Colab ! Amazing work !
I assume the method works also on a T4 since you shared a colab demo, would you mind adding simple tests? 
You can simply copy paste the tests from AWQ: https://github.com/huggingface/transformers/blob/main/tests/quantization/autoawq/test_awq.py and simply have Config tests and very simple model tests that tests that the model has been successfully converted to Aqlm format + a generation test. I would also add a simple test to make sure the model loads well on CPU 
Can you also share some insights on generation speed for CPU & GPU ? 🙏

younesbelkada reviewed (commented):
One last thing; could you add installation instructions on our testing dockerfile: https://github.com/huggingface/transformers/blob/main/docker/transformers-all-latest-gpu/Dockerfile

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28928). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

younesbelkada commented:
cc @oobabooga this might be of your interest !

Commit:
Readme update

Commit:
is_aqlm_available

Commit:
Simpler link in docs

Commit:
Test TODO real reference

Commit:
init _import_structure fix

Commit:
AqlmConfig autodoc

BlackSamorez commented:
I updated the docked recipe and added tests, but they are skipped because `aqlm` is not installed in the testing environment.

younesbelkada commented on src/transformers/quantizers/quantizer_aqlm.py:
```diff
@@ -0,0 +1,144 @@
+        return True
+
+
+def _replace_with_aqlm_linear(
```
you can move this method and make it public under `integrations/aqlm.py` and import locally the method inside `_process_model_before_weight_loading`

BlackSamorez replied:
Done

younesbelkada commented on tests/quantization/aqlm/test_aqlm.py:
```diff
@@ -0,0 +1,182 @@
+
+
+@require_torch_gpu
+class AwqConfigTest(unittest.TestCase):
```
```suggestion
class AqlmConfigTest(unittest.TestCase):
```

BlackSamorez replied:
Done

younesbelkada reviewed (commented):
Thanks a lot ! I left two minor comments ! Can you also run `make fixup` ? This will redirect you to run `make fix-copies` which should fix the tests !

Commit:
integration aqlm

Commit:
integrations in tests

Commit:
docstring fix

Commit:
legacy typing

Commit:
Less typings

BlackSamorez commented:
The tests are still getting skipped. Is putting `aqlm` in the dockerfile enough for the testing environment to get it?

Commit:
More kernels information

Commit:
Performance -> Accuracy

Commit:
correct tests

Commit:
remoced multi-gpu test

BlackSamorez commented:
I'm pretty sure tests failing has nothing to do with my PR

younesbelkada reviewed (approved):
Looks very clean ! Thanks so much for the integration !
As discussed with @ArthurZucker offline, we could leverage `# Copied from` on the tests but this is clearly not a blocker for me and we can merge as is! Looking forward to the integration !

ArthurZucker commented on src/transformers/utils/quantization_config.py:
```diff
@@ -731,3 +732,63 @@ def get_loading_attributes(self):
+            Number of codebooks for the Additive Quantization procedure.
+        nbits_per_codebook (`int`, *optional*, defaults to 16):
+            Number of bits encoding a single codebook vector. Codebooks size is 2**nbits_per_codebook.
+        linear_weights_not_to_quantize (`Optional[List[str]]`, *optional*, default to `None`):
```
```suggestion
        linear_weights_not_to_quantize (`Optional[List[str]]`, *optional*):
```

ArthurZucker reviewed (approved):
LGTM thanks for the clean PR 🤗

younesbelkada commented on docs/source/en/quantization.md:
```diff
@@ -26,6 +26,34 @@ Interested in adding a new quantization method to Transformers? Read the [HfQuan
+
+Additive Quantization of Language Models ([AQLM](https://arxiv.org/abs/2401.06118)) is a Large Language Models compression method. It quantizes multiple weights together and take advantage of interdependencies between them. AQLM represents groups of 8-16 weights as a sum of multiple vector codes.
+
+Inference support for AQLM is realised in the `aqlm` library. Make sure to install it to run the models:
```
```suggestion
Inference support for AQLM is realised in the `aqlm` library. Make sure to install it to run the models (note aqlm works only with python>=3.10):
```

younesbelkada reviewed (approved):
I also left one nit ! I tested it and seems to work only for python >=3.10

younesbelkada commented:
@BlackSamorez on a google colab env the inference script works great, however on my VM, on a python 3.10 env with latest torch + cuda11.8 I constantly get:
```bash
Traceback (most recent call last):
  File "/home/younes_huggingface_co/code/transformers/scratch.py", line 11, in <module>
    output = quantized_model.generate(tokenizer("", return_tensors="pt")["input_ids"].cuda(), max_new_tokens=10)
  File "/home/younes_huggingface_co/miniconda3/envs/aqlm/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/younes_huggingface_co/code/transformers/src/transformers/generation/utils.py", line 1495, in generate
    return self.greedy_search(
  File "/home/younes_huggingface_co/code/transformers/src/transformers/generation/utils.py", line 2366, in greedy_search
    next_tokens = torch.argmax(next_tokens_scores, dim=-1)
RuntimeError: CUDA error: device kernel image is invalid
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```
Do you have an idea what might be wrong here?

younesbelkada commented:
@BlackSamorez on a google colab env the inference script works great, however on my VM, on a python 3.10 env with latest torch + cuda11.8 I constantly get:
```bash
Traceback (most recent call last):
  File "/transformers/scratch.py", line 11, in <module>
    output = quantized_model.generate(tokenizer("", return_tensors="pt")["input_ids"].cuda(), max_new_tokens=10)
  File "/miniconda3/envs/aqlm/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/transformers/src/transformers/generation/utils.py", line 1495, in generate
    return self.greedy_search(
  File "/transformers/src/transformers/generation/utils.py", line 2366, in greedy_search
    next_tokens = torch.argmax(next_tokens_scores, dim=-1)
RuntimeError: CUDA error: device kernel image is invalid
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```
Do you have an idea what might be wrong here?

younesbelkada commented:
The only difference I see between the colab instance and mine is the CUDA version, I'll update it to 12.1 and loop back here

Commit:
Update docs/source/en/quantization.md

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

Commit:
Update src/transformers/utils/quantization_config.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

BlackSamorez commented:
@younesbelkada 
The kernel is compiled in runtime with 
```python
import os
from typing import Optional

import torch
from torch.utils.cpp_extension import load

CUDA_FOLDER = os.path.dirname(os.path.abspath(__file__))
CUDA_KERNEL = load(
    name="codebook_cuda",
    sources=[os.path.join(CUDA_FOLDER, "cuda_kernel.cpp"), os.path.join(CUDA_FOLDER, "cuda_kernel.cu")],
)
```

Maybe your `nvcc` is sourced from an incorrect cuda installment. I'm not really sure how to test it. Maybe you could try specifying an `nvcc` path somehow with an environmental variable.
I'll try to reproduce it as well.

BlackSamorez commented:
CUDA 11.8 seems to work fine on my machine on an a100 GPU.

Commit:
Brought back multi-gpu tests

Commit:
Merge branch 'aqlm' of github.com:BlackSamorez/transformers into aqlm

BlackSamorez commented:
FYI: I've released `aqlm` version `1.0.1` where I added device locks to fix CUDA errors when running in the multi-gpu setup. I've added the corresponding tests similar to `autoawq` ones

SunMarc reviewed (approved):
Thanks for the integration !

Commit:
Update src/transformers/integrations/aqlm.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

Commit:
Update tests/quantization/aqlm_integration/test_aqlm.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

BlackSamorez commented:
Looks like some network error occured

younesbelkada reviewed (approved):
Thanks a lot @BlackSamorez !

ArthurZucker commented:
🤗 🚀

younesbelkada commented:
Hi @BlackSamorez ! 
Thanks again for your great work ! I was wondering if you could update the installation cell on the shared notebook to install transformers from source instead of your fork - that way we could catch potential bugs in the future before the release 🙏

BlackSamorez commented:
@younesbelkada 
Looks like [this commit](https://github.com/huggingface/transformers/commit/164bdef8cc5143a0766cee448e97166682a722b1) outside of the PR broke something.
```python
AttributeError                            Traceback (most recent call last)

[<ipython-input-2-68b1b199d504>](https://localhost:8080/#) in <cell line: 3>()
      1 from transformers import AutoTokenizer, AutoModelForCausalLM
      2 
----> 3 quantized_model = AutoModelForCausalLM.from_pretrained(
      4     "BlackSamorez/Mixtral-8x7b-AQLM-2Bit-1x16-hf-test-dispatch",
      5     torch_dtype="auto", device_map="auto", low_cpu_mem_usage=True,

4 frames

[/usr/local/lib/python3.10/dist-packages/transformers/models/auto/auto_factory.py](https://localhost:8080/#) in from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs)
    565         elif type(config) in cls._model_mapping.keys():
    566             model_class = _get_model_class(config, cls._model_mapping)
--> 567             return model_class.from_pretrained(
    568                 pretrained_model_name_or_path, *model_args, config=config, **hub_kwargs, **kwargs
    569             )

[/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py](https://localhost:8080/#) in from_pretrained(cls, pretrained_model_name_or_path, config, cache_dir, ignore_mismatched_sizes, force_download, local_files_only, token, revision, use_safetensors, *model_args, **kwargs)
   3561 
   3562         if hf_quantizer is not None:
-> 3563             hf_quantizer.postprocess_model(model)
   3564             model.hf_quantizer = hf_quantizer
   3565 

[/usr/local/lib/python3.10/dist-packages/transformers/quantizers/base.py](https://localhost:8080/#) in postprocess_model(self, model, **kwargs)
    177                 The keyword arguments that are passed along `_process_model_after_weight_loading`.
    178         """
--> 179         return self._process_model_after_weight_loading(model, **kwargs)
    180 
    181     @abstractmethod

[/usr/local/lib/python3.10/dist-packages/transformers/quantizers/quantizer_aqlm.py](https://localhost:8080/#) in _process_model_after_weight_loading(self, model, **kwargs)
     78 
     79     def _process_model_after_weight_loading(self, model: "PreTrainedModel", **kwargs):
---> 80         model._is_quantized_training_enabled = False
     81         return model
     82 

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in __setattr__(self, name, value)
   1745                     buffers[name] = value
   1746                 else:
-> 1747                     super().__setattr__(name, value)
   1748 
   1749     def __delattr__(self, name):

AttributeError: can't set attribute '_is_quantized_training_enabled'
```
